[TypeChecker] Allow opaque to have multiple underlying types based on availability

xedin opened the pull request:
Allow opaque type declarations to have multiple unique underlying
types if all such types are conditional on availability via
`if #available(...)` and there is only one universally available
substitution.  Neither nesting of availability conditions nor other
dynamic conditions are allowed.

For example:

```swift
protocol P {}

@available(iOS 31.337, *)
struct X : P {
}

struct Y : P {
}

func test() -> some P {
  if #available(iOS 31.337, *) {
    // ... some computation ...
    return X()
  }

  return Y()
 }
 ```

 Allows `some P` to refer to `X` instead of `Y` on `iOS 31.337`.

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

xedin committed:
[AST] Augment OpaqueTypeDecl to store conditionally available alternatives

In preparation to support cases like this:

```
func test() -> some P {
  if #available(...) {
    return X()
  }

  return Y()
}
```

where both `X` and `Y` conform to `P` but are different types.

xedin committed:
[AST] NFC: Rename underlying type accessors of `OpaqueTypeDecl`

Adds a `Unique` component to the accessor names to draw the
distinction between unique and conditionally available substitutions.

LucianoPAlmeida commented on test/type/opaque_with_conditional_availability.swift:
```diff
@@ -0,0 +1,108 @@
+// RUN: %target-swift-frontend -target x86_64-apple-macosx10.15 -typecheck -verify %s
```
Would this have the same problem as the `spi-available-inline` and fail running locally on M1 macs? Because swift-frontend target will try load x86_64 stdlib and fail, so should we use dynamic target as well `%target-cpu-apple-macosx10.15`?

xedin replied:
Yeah, I guess that's reasonable to do...

xedin committed:
[TypeChecker] Allow opaque to have multiple underlying types based on availability

Allow opaque type declarations to have multiple unique underlying
types if all such types are conditional on availability via
`if #available(...)` and there is only one universally available
substitution. Neither nesting of availability conditions nor other
dynamic conditions are allowed.

For example:

```swift
protocol P {}

@available(iOS 31.337, *)
struct X : P {
}

struct Y : P {
}

func test() -> some P {
  if #available(iOS 31.337, *) {
    // ... some computation ...
    return X()
  }

  return Y()
}
```

Allows `some P` to refer to `X` instead of `Y` on `iOS 31.337`.

xedin commented:
@swift-ci please test

xedin committed:
[AST] OpaqueDecl: Simplify storage of conditionally available alternatives

Instead of using a special type with a set of trailing objects, let's
just re-allocate array of substitutions into a permanent storage directly.

xedin commented:
@swift-ci please test
